sstables_loader: report progress with the unit of batch

tchaikov opened the pull request:
Previously, restore progress was tracked using individual sstables as the unit of measurement. This approach introduced inaccuracies for tables distributed with tablets, where:
- An sstable spanning multiple tablets could be counted multiple times
- Progress reporting could become misleading (e.g., showing "40" progress for a table with 10 sstables)

This change introduces a more robust progress tracking method:
- Use "batch" as the unit of progress instead of individual sstables
- Stream sstables for each tablet separately, handling both partially and fully contained sstables
- Calculate progress based on the total number of sstables being streamed
- Skip tablet IDs with no owned tokens

For vnode-distributed tables, the number of "batches" directly corresponds to the number of sstables, ensuring:
- Consistent progress reporting across different table distribution models
- Simplified implementation
- Accurate representation of restore progress

The new approach provides a more reliable and uniform method of tracking restoration progress across different table distribution strategies.

Fixes scylladb/scylladb#21816

---

the feature being addressed by this fix is still experimental, hence no need to backport.

tchaikov commented on sstables_loader.cc:
```diff
@@ -257,13 +264,25 @@ future<> tablet_sstable_streamer::stream(std::function<void(unsigned)> on_stream
+            double progress_unit = 1. / _tablet_map.tablet_count();
+            size_t num_sstables = sstables_fully_contained.size() + sstables_partially_contained.size();
+            if (num_sstables > 0) {
+                on_streamed = [unit = progress_unit / num_sstables, on_streamed = std::move(on_streamed)] (double processed) {
```
i am not sure this helps with the readability. probably we can wait for more inputs.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13708/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13708/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/590/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13268/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5260/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2154/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13708/)
#### Build Details:
- Duration: 8 hr 29 min
- Builder: i-091eca43af7ca90fc (m5ad.8xlarge)

tchaikov commented:
@xemul hi Pavel, could you please help review this change?

xemul commented:
> An sstable spanning multiple tablets could be counted multiple times

How can this happen? Sstables are not allowed to cross tablet boundaries

xemul commented on sstables_loader.cc:
```diff
@@ -518,8 +554,15 @@ future<> sstables_loader::download_task_impl::run() {
+            } else {
+                // negative for total
+                processed.second = -n;
+            }
```
That's extremely confusing to use update value sign to indicate different pars of the progress being updated.
Can we replace `std::function<void(unsigned)> on_streamed` with `lw_shared_ptr<struct { unsigned progress; unsigned total; }> progress` and update its members in streamer methods directly?

tchaikov replied:
> That's extremely confusing to use update value sign to indicate different pars of the progress being updated. 

agreed.  neither am i in favor of this design. 

> Can we replace `std::function<void(unsigned)> on_streamed` with `lw_shared_ptr<struct { unsigned progress; unsigned total; }> progress` and update its members in streamer methods directly?

please note, i am using `function<void(double)>` in the latest revision. because in order to use a single interface of `sstable_streamer::stream_sstables()` shared by restore processes of both tables with tablets and vnodes, we cannot use "unsigned". because,  in `tablet_sstable_streamer::stream()`, we stream sstables on a per-tablet basis -- we iterate through the tablets' token ranges, and stream the sstables providing each tablet in batch size of a single tablet, and we need to advance the progress with a finer granularity of 16 sstables in each batch.

probably using the smart pointer is to be defensive and probably "safer", but i think, the coroutine or sub coroutine spawned by `loader.load_and_stream()` cannot outlive its caller. we can we just pass a raw pointer to the callee. 

but still, i don't think it's a good idea even we use a raw pointer, which exposes the internals of the caller to callee. so i will use `std::function<void(double, double)>`.

xemul replied:
>  i don't think it's a good idea even we use a raw pointer, which exposes the internals of the caller to callee. so i will use std::function<void(double, double)>.

I'm fine with using a callable to update progress, but the `std::function<void(double, double)>` is even more confusing :( How about introducing a `download_progress_monitor` class with two methods `.set_total()` (or `.update_total()` and `.update_progress()` (or `.add_progress()`)?

tchaikov replied:
hmm, it is confusing after rereading the code. i added two classes for tracking the progress, and accompanied them with some comment.

tchaikov commented:
> > An sstable spanning multiple tablets could be counted multiple times
> 
> How can this happen? Sstables are not allowed to cross tablet boundaries

i had a hard time understanding https://github.com/scylladb/scylladb/blob/5880a1b90b45664f309a156116f3cf8f7d39d6e5/sstables_loader.cc#L248-L257 

i think we are doing something like (in Python pseudo code):
```py
sst_fully_contained = []
sst_partially_contained = []

sstables = iter(self.sstables)
for tablet_id in self.tablet_map.tablet_ids:
  tablet_range = self.tablet_map.token_ranges[tablet_id]
  for sst in sstables:
    #                             V
    #   |<--- tablet 1 -->|<--- tablet 2 --->|
    #     |<- sst ->|
    if sst.last_token < tablet_range:
      continue
    sst_range = Range(sst.first_token, sst.last_token)
    #                             V
    #   |<--- tablet 1 -->|<--- tablet 2 --->|
    #               |<- sst ->|
    if not tablet_range.overlap_with(sst_range):
      break
    if tablet_range.contains(sst_token):
      #                             V
      #   |<--- tablet 1 -->|<--- tablet 2 --->|
      #                     |<--- sst -->|
      sst_fully_contained.append(sst)
    else:
      #                             V
      #   |<--- tablet 1 -->|<--- tablet 2 --->|
      #                   |<--- sst -->|
      sst_partially_contained.append(sst)
  await stream_sstables(sst_fully_contained)
  await stream_sstables(sst_partially_contained)
```

> Sstables are not allowed to cross tablet boundaries

based on the analysis, i came to a different conclusion.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13868/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13868/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
#### Failed Tests (30/180):
- [test_simple_backup_and_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.1)
- [test_simple_backup_and_restore.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.10)
- [test_simple_backup_and_restore.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.2)
- [test_simple_backup_and_restore.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.3)
- [test_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.4)
- [test_simple_backup_and_restore.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.5)
- [test_simple_backup_and_restore.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.6)
- [test_simple_backup_and_restore.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.7)
- [test_simple_backup_and_restore.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.8)
- [test_simple_backup_and_restore.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.9)
#### Build Details:
- Duration: 29 min
- Builder: spider5.cloudius-systems.com

tchaikov commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13868/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13868/) **The following new/updated tests ran 100 times for each mode:** 🔹 object_store/test_backup
> 
> #### Failed Tests (30/180):
> * [test_simple_backup_and_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.1)
> * [test_simple_backup_and_restore.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_10) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.10)
> * [test_simple_backup_and_restore.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.2)
> * [test_simple_backup_and_restore.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.3)
> * [test_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.4)
> * [test_simple_backup_and_restore.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.5)
> * [test_simple_backup_and_restore.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.6)
> * [test_simple_backup_and_restore.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_7) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.7)
> * [test_simple_backup_and_restore.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_8) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.8)
> * [test_simple_backup_and_restore.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13868/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_9) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.9)
> 
> #### Build Details:
> * Duration: 29 min
> * Builder: spider5.cloudius-systems.com

i introduced a regression, looking..

tchaikov commented:
v2:

- instead of using negative value for passing "total" and using positive value for passing "progress", using two different parameters for passing "total" and "progress".
-  change `std::function<void(double)>` to `std::function<void(double,double>)`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13889/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13889/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5460/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2341/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13471/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/141/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/739/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13889/)
#### Failed Tests (5/37950):
- [test_backup_restore_with_nodetool](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest_with_topology_changes___test_backup_restore_with_nodetool) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_restore_with_nodetool)
- [test_restore_into_different_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest_with_topology_changes___test_restore_into_different_table) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_into_different_table)
- [test_backup_restore_with_nodetool](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest___test_backup_restore_with_nodetool) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_restore_with_nodetool)
- [test_restore_into_different_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest___test_restore_into_different_table) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_into_different_table)
- [memtable_flush_period](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/boost/memtable_test/memtable_flush_period) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+memtable_flush_period)
#### Build Details:
- Duration: 7 hr 15 min
- Builder: i-0129dec23a7dfc432 (m5ad.12xlarge)

tchaikov commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13889/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13889/) **The following new/updated tests ran 100 times for each mode:** 🔹 object_store/test_backup ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5460/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2341/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13471/) ✅ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/141/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/739/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13889/)
> #### Failed Tests (5/37950):
> 
>     * [test_backup_restore_with_nodetool](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest_with_topology_changes___test_backup_restore_with_nodetool) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_restore_with_nodetool)
> 
>     * [test_restore_into_different_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest_with_topology_changes___test_restore_into_different_table) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_into_different_table)
> 
>     * [test_backup_restore_with_nodetool](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest___test_backup_restore_with_nodetool) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_restore_with_nodetool)
> 
>     * [test_restore_into_different_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/backup_restore_object_storage_tests/TestBackupRestoreObjectStorage/Tests___dtest___test_restore_into_different_table) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_into_different_table)
> 
>     * [memtable_flush_period](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13889/testReport/junit/boost/memtable_test/memtable_flush_period) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+memtable_flush_period)
> 
> 
> #### Build Details:
> 
>     * Duration: 7 hr 15 min
> 
>     * Builder: i-0129dec23a7dfc432 (m5ad.12xlarge)

these test failures were caused by the change in previous revision of this change. should have been fixed in the latest revision.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13924/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13924/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/770/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13507/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5495/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2378/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/172/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13924/)
#### Failed Tests (4/38131):
- [test_mv_topology_change.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13924/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.2)
- [test_replace_with_no_normal_token_owners_in_dc[False].release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13924/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_False__release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[False].release.3)
- [test_random_failures[stop_before_becoming_raft_voter-kill_coordinator_node].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13924/testReport/junit/topology_random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_before_becoming_raft_voter_kill_coordinator_node__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_before_becoming_raft_voter-kill_coordinator_node].debug.2)
- [test_mv_topology_change.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13924/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.3)
#### Build Details:
- Duration: 8 hr 26 min
- Builder: i-00438e4cb7e87d20a (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13984/)
#### Build Details:
- Duration: 19 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13985/)
#### Build Details:
- Duration: 10 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13986/)
#### Build Details:
- Duration: 12 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13988/)
#### Build Details:
- Duration: 13 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13989/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13989/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
#### Failed Tests (30/210):
- [test_simple_backup_and_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.1)
- [test_simple_backup_and_restore.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.10)
- [test_simple_backup_and_restore.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.2)
- [test_simple_backup_and_restore.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.3)
- [test_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.4)
- [test_simple_backup_and_restore.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.5)
- [test_simple_backup_and_restore.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.6)
- [test_simple_backup_and_restore.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.7)
- [test_simple_backup_and_restore.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.8)
- [test_simple_backup_and_restore.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.9)
#### Build Details:
- Duration: 27 min
- Builder: spider2.cloudius-systems.com

tchaikov commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13989/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13989/) **The following new/updated tests ran 100 times for each mode:** 🔹 object_store/test_backup
> #### Failed Tests (30/210):
> 
>     * [test_simple_backup_and_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.1)
> 
>     * [test_simple_backup_and_restore.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_10) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.10)
> 
>     * [test_simple_backup_and_restore.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.2)
> 
>     * [test_simple_backup_and_restore.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.3)
> 
>     * [test_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.4)
> 
>     * [test_simple_backup_and_restore.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.5)
> 
>     * [test_simple_backup_and_restore.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.6)
> 
>     * [test_simple_backup_and_restore.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_7) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.7)
> 
>     * [test_simple_backup_and_restore.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_8) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.8)
> 
>     * [test_simple_backup_and_restore.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_9) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.9)
> 
> 
> #### Build Details:
> 
>     * Duration: 27 min
> 
>     * Builder: spider2.cloudius-systems.com

this was a failure due to a buggy rebase. it's fixed in the latest revision.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13993/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13993/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/817/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13560/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5546/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2429/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/219/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13993/)
#### Build Details:
- Duration: 4 hr 2 min
- Builder: spider4.cloudius-systems.com

denesb commented:
> > An sstable spanning multiple tablets could be counted multiple times
> 
> How can this happen? Sstables are not allowed to cross tablet boundaries

Sstables uploaded via `nodetool refresh` are not guaranteed to confirm to current tablet boundaries. Even in the case where ScyllaDB produced the original sstables and they are uploaded to the same or another cluster.

tchaikov commented:
@xemul hi Pavel, could you please take another look?

tchaikov commented:
> > > An sstable spanning multiple tablets could be counted multiple times
> > 
> > 
> > How can this happen? Sstables are not allowed to cross tablet boundaries
> 
> Sstables uploaded via `nodetool refresh` are not guaranteed to confirm to current tablet boundaries. Even in the case where ScyllaDB produced the original sstables and they are uploaded to the same or another cluster.

thank you Botond for explaining this, i didn't know the rationales behind the code.

xemul commented on sstables_loader.cc:
```diff
@@ -502,13 +525,30 @@ class sstables_loader::download_task_impl : public tasks::task_manager::task::im
+                if (result.total == 0.) {
+                    // total number of batches can be tracked on multiple shards,
+                    // let's keep the first non-zero one.
+                    result.total = p.total;
```
Why is that so? E.g. the `sstable_streamer::stream()` calls progress update like `on_stream(0, _sstables.size())` which means that local progress' .total is updated with local _sstables.size(), respectively total number of sstables should be summed on all shards, not just picked from "random" one

tchaikov replied:
thank you for pointing this out indeed, we split the sstables by shard, so we should sum them up, instead of picking a "random" one.

xemul commented on sstables_loader.cc:
```diff
@@ -258,14 +265,26 @@ future<> tablet_sstable_streamer::stream(std::function<void(unsigned)> on_stream
+            size_t num_sstables = sstables_fully_contained.size() + sstables_partially_contained.size();
+            if (num_sstables > 0) {
+                on_streamed = [unit = progress_unit / num_sstables, on_streamed = std::move(on_streamed)] (double processed, double) {
+                    std::invoke(on_streamed, processed * unit, 0.);
```
This updates progress with the value of
```
processed * unit = processed * progress_unit / num_sstables = processed * (1 / _tablet_map.tablet_count()) / all_sstables.size() = processed / (_tablet_map.tablet_count() * all_sstables.size())
```

Next, `tablet_sstable_streamer::stream()` sets "total" to be `_tablet_map.tablet_count()`

Finally, the implementation of on_streamed() does this:

```
   progress.processed += processed;
   progress.total = std::max(progress.total, total);
```

In the end of the day `progress.total = _tablet_map.tablet_count()` and `progress.processed = all_sstables.size() * _tablet_map.count()`

Please, correct my understanding

tchaikov replied:
> This updates progress with the value of
> 
> ```
> processed * unit = processed * progress_unit / num_sstables = processed * (1 / _tablet_map.tablet_count()) / all_sstables.size() = processed / (_tablet_map.tablet_count() * all_sstables.size())
> ```
> 
> Next, `tablet_sstable_streamer::stream()` sets "total" to be `_tablet_map.tablet_count()`
> 
> Finally, the implementation of on_streamed() does this:
> 
> ```
>    progress.processed += processed;
>    progress.total = std::max(progress.total, total);
> ```
> 
> In the end of the day `progress.total = _tablet_map.tablet_count()` and `progress.processed = all_sstables.size() * _tablet_map.count()`
> 

i think the bug in previous revision is that, 

1. i should not advance the progress in the inner loop by "batch_sst_nr / sstables" -- i should have just call `on_stream(batch_sst_nr)", as i've already divided it by "sstables_partially_contained.size() + sstables_fully_contained.size()"
2. i should not divide the unit by `_tablet_map.tablet_count()`. because the unit should be "1". and the sum of all the units are ` _tablet_map.tablet_count()`

anyway, the math is confusing. so i reworked it by introducing another progress class for tracking the per-tablet progress.

tchaikov commented:
v3:

- rebased against master HEAD
- instead of using a callable, introduced two dedicated class for monitoring the progress of restoring
- corrected the calculation of progress in `tablet_sstable_streamer::stream()`. previously, i divided the progress by num of tablets and then by sstables by mistake.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14105/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14105/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
#### Failed Tests (70/390):
- [test_restore_with_streaming_scopes[topology0].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.1)
- [test_restore_with_streaming_scopes[topology0].debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.10)
- [test_restore_with_streaming_scopes[topology0].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.2)
- [test_restore_with_streaming_scopes[topology0].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.3)
- [test_restore_with_streaming_scopes[topology0].debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.4)
- [test_restore_with_streaming_scopes[topology0].debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.5)
- [test_restore_with_streaming_scopes[topology0].debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.6)
- [test_restore_with_streaming_scopes[topology0].debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.7)
- [test_restore_with_streaming_scopes[topology0].debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.8)
- [test_restore_with_streaming_scopes[topology0].debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14105/testReport/junit/object_store/test_backup/test_restore_with_streaming_scopes_topology0__debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_with_streaming_scopes[topology0].debug.9)
#### Build Details:
- Duration: 35 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14110/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14110/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
#### Failed Tests (30/390):
- [test_abort_simple_backup_and_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.1)
- [test_abort_simple_backup_and_restore.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.10)
- [test_abort_simple_backup_and_restore.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.2)
- [test_abort_simple_backup_and_restore.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.3)
- [test_abort_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.4)
- [test_abort_simple_backup_and_restore.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.5)
- [test_abort_simple_backup_and_restore.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.6)
- [test_abort_simple_backup_and_restore.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.7)
- [test_abort_simple_backup_and_restore.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.8)
- [test_abort_simple_backup_and_restore.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.9)
#### Build Details:
- Duration: 35 min
- Builder: spider2.cloudius-systems.com

tchaikov commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14110/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14110/) **The following new/updated tests ran 100 times for each mode:** 🔹 object_store/test_backup
> #### Failed Tests (30/390):
> 
>     * [test_abort_simple_backup_and_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.1)
> 
>     * [test_abort_simple_backup_and_restore.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_10) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.10)
> 
>     * [test_abort_simple_backup_and_restore.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.2)
> 
>     * [test_abort_simple_backup_and_restore.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.3)
> 
>     * [test_abort_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.4)
> 
>     * [test_abort_simple_backup_and_restore.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.5)
> 
>     * [test_abort_simple_backup_and_restore.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.6)
> 
>     * [test_abort_simple_backup_and_restore.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_7) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.7)
> 
>     * [test_abort_simple_backup_and_restore.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_8) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.8)
> 
>     * [test_abort_simple_backup_and_restore.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14110/testReport/junit/object_store/test_backup/test_abort_simple_backup_and_restore_debug_9) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_abort_simple_backup_and_restore.debug.9)
> 
> 
> #### Build Details:
> 
>     * Duration: 35 min
> 
>     * Builder: spider2.cloudius-systems.com

this failure and the previous one have been fixed in the latest revision.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14112/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14112/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5651/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13664/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2537/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14112/)
:x: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/909/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/313/)
#### Failed Tests (367/40478):
- [test_no_data_resurrection_or_loss_after_restore](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest___test_no_data_resurrection_or_loss_after_restore) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_data_resurrection_or_loss_after_restore)
- [test_failure_durring_restore_no_corrupt_data](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_failure_durring_restore_no_corrupt_data) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_failure_durring_restore_no_corrupt_data)
- [test_failure_durring_restore_no_corrupt_data](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_failure_durring_restore_no_corrupt_data_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_failure_durring_restore_no_corrupt_data)
- [test_replay_restore_no_additional_data](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_replay_restore_no_additional_data) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_restore_no_additional_data)
- [test_replay_restore_no_additional_data](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_replay_restore_no_additional_data_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_restore_no_additional_data)
- [test_restore_snapshot_using_different_smp_setting](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_different_smp_setting) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_different_smp_setting)
- [test_restore_snapshot_using_different_smp_setting](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_different_smp_setting_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_different_smp_setting)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_using_old_token_ownership_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_restore_snapshot_from_cassandra](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14112/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_tablets___test_restore_snapshot_from_cassandra) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_from_cassandra)
#### Build Details:
- Duration: 6 hr 24 min
- Builder: spider2.cloudius-systems.com

tchaikov commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13989/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13989/) **The following new/updated tests ran 100 times for each mode:** 🔹 object_store/test_backup
> #### Failed Tests (30/210):
> 
>     * [test_simple_backup_and_restore.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.1)
> 
>     * [test_simple_backup_and_restore.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_10) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.10)
> 
>     * [test_simple_backup_and_restore.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.2)
> 
>     * [test_simple_backup_and_restore.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.3)
> 
>     * [test_simple_backup_and_restore.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.4)
> 
>     * [test_simple_backup_and_restore.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.5)
> 
>     * [test_simple_backup_and_restore.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.6)
> 
>     * [test_simple_backup_and_restore.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_7) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.7)
> 
>     * [test_simple_backup_and_restore.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_8) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.8)
> 
>     * [test_simple_backup_and_restore.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13989/testReport/junit/object_store/test_backup/test_simple_backup_and_restore_debug_9) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_backup_and_restore.debug.9)
> 
> 

these test failures were caused by the changes in the previous revisions of this fix. should have been fixed in the latest revision.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14115/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14115/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5653/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13666/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2539/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/912/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/316/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14115/)
#### Build Details:
- Duration: 5 hr 19 min
- Builder: spider4.cloudius-systems.com

tchaikov commented:
@xemul hi Pavel, could you help re-review this change?

avikivity commented:
> Use "batch" as the unit of progress instead of individual sstables

What is a "batch" unit?

Can't we use bytes?

tchaikov commented:
> > Use "batch" as the unit of progress instead of individual sstables
> 
> What is a "batch" unit?

i improve the commit message to explain "batch".

> 
> Can't we use bytes?

Using bytes as a progress unit isn't feasible due to how `sstable_streamer::stream_sstables()` processes mutations:

1. We can't accurately pre-calculate the total mutation size because:
   - Mutations are read and streamed in batches
   - The actual mutation data size may differ from the raw sstable size
   - Tombstones affect the final mutation count

2. Tracking progress in bytes would require complex byte counting for each streamed mutation

3. The batch-based approach provides a simpler, more reliable metric while 
   avoiding the complexity and potential inaccuracies of byte-level tracking.

4. More importantly, the total size of source sstables wouldn't match the bytes 
   of mutations sent to target nodes.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14192/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14192/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/964/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2594/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5707/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13720/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/368/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14192/)
#### Build Details:
- Duration: 5 hr 3 min
- Builder: spider8.cloudius-systems.com

xemul reviewed (approved):
Looks OK

tchaikov commented:
@scylladb/scylla-maint hello maintainers, could you help queue this change?
